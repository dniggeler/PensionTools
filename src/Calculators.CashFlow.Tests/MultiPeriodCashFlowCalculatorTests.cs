using System.Collections.Generic;
using System.Threading.Tasks;
using Calculators.CashFlow.Models;
using PensionCoach.Tools.CommonTypes;
using Snapshooter.Xunit;
using Xunit;

namespace Calculators.CashFlow.Tests
{
    [Trait("Higher Level Calculators", "Multi-Period Calculator")]
    public class MultiPeriodCashFlowCalculatorTests : IClassFixture<CashFlowFixture<IMultiPeriodCashFlowCalculator>>
    {
        private readonly CashFlowFixture<IMultiPeriodCashFlowCalculator> _fixture;

        public MultiPeriodCashFlowCalculatorTests(CashFlowFixture<IMultiPeriodCashFlowCalculator> fixture)
        {
            _fixture = fixture;
        }

        [Fact(DisplayName = "Salary and Wealth Projection")]
        public async Task Calculate_MultiPeriod_Simulation_With_Salary_And_Wealth()
        {
            // given
            int startingYear = 2021;
            int numberOfPeriods = 10;
            int municipalityId = 261; // Zurich
            Canton canton = Canton.ZH;
            MultiPeriodCalculatorPerson person = new()
            {
                CivilStatus = CivilStatus.Married,
                Name = "Unit Test",
                Canton = canton,
                MunicipalityId = municipalityId,
                Income = 100_000,
                Wealth = 500_000,
                NumberOfChildren = 0,
                PartnerReligiousGroupType = ReligiousGroupType.Other,
                ReligiousGroupType = ReligiousGroupType.Other
            };

            List<GenericCashFlowDefinition> cashFlowDefinitions = new();
            //{
            //    new GenericCashFlowDefinition
            //    {
            //        NetGrowthRate = 0,
            //        Name = "Test",
            //        InitialAmount = 6883,
            //        RecurringAmount = (6883, FrequencyType.Yearly),
            //        Flow = (AccountType.Income, AccountType.CapitalBenefits),
            //        InvestmentPeriod = (2021, 10),
            //        IsTaxable = true,
            //        TaxType = TaxType.Income
            //    }
            //};

            Dictionary<AccountType, decimal> initialFunds = new Dictionary<AccountType, decimal>();

            // when
            var result = await _fixture.Service.CalculateAsync(
                startingYear, numberOfPeriods, person, cashFlowDefinitions, initialFunds);

            // then
            Snapshot.Match(result);
        }
    }
}
