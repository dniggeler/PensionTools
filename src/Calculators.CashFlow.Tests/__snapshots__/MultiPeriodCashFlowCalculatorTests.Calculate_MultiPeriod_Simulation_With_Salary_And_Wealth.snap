{
  "StartingYear": 2021,
  "NumberOfPeriods": 10,
  "Accounts": [
    {
      "Year": 2021,
      "Amount": 516730.865,
      "AccountType": "Wealth"
    },
    {
      "Year": 2022,
      "Amount": 534041.3250,
      "AccountType": "Wealth"
    },
    {
      "Year": 2023,
      "Amount": 551429.585000,
      "AccountType": "Wealth"
    },
    {
      "Year": 2024,
      "Amount": 568898.67500000,
      "AccountType": "Wealth"
    },
    {
      "Year": 2025,
      "Amount": 586445.6553000000,
      "AccountType": "Wealth"
    },
    {
      "Year": 2026,
      "Amount": 604046.096803000000,
      "AccountType": "Wealth"
    },
    {
      "Year": 2027,
      "Amount": 621726.64132103000000,
      "AccountType": "Wealth"
    },
    {
      "Year": 2028,
      "Amount": 639466.9218842403000000,
      "AccountType": "Wealth"
    },
    {
      "Year": 2029,
      "Amount": 657293.643053082703000000,
      "AccountType": "Wealth"
    },
    {
      "Year": 2030,
      "Amount": 675186.50123361353003000000,
      "AccountType": "Wealth"
    }
  ]
}
